LCN component and light platform

alengwenus opened the pull request:
## Description:
Implementation of the [LCN](https://www.lcn.eu) (Local Control Network). LCN is a well known and popular home automation system from Germany with a broad range of devices, sensors, actuators, and much more.
This PR covers the base component and light platform.

**Related issue (if applicable):**

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7599

## Example entry for `configuration.yaml` (if applicable):
```yaml
lcn:
  connections:
    - name: myhome
      host: 192.168.2.41
      port: 4114
      username: !secret lcn_username
      password: !secret lcn_password

  lights:
    - name: Living room
      address: myhome.0.7
      output: output1
      dimmable: true
      transition: 5
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

alengwenus committed:
Initial commit of LCN component and light platform

fabaff commented on homeassistant/components/lcn/__init__.py:
```diff
@@ -0,0 +1,118 @@
+"""
+
+import logging
+import voluptuous as vol
```
Please keep the imports grouped.

alengwenus replied:
Done.

fabaff commented on homeassistant/components/lcn/__init__.py:
```diff
@@ -0,0 +1,118 @@
+
+CONNECTION_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_PORT): cv.positive_int,
```
We have `cv.port` to check if it's a valid port number.

alengwenus replied:
Replaced.

fabaff commented on homeassistant/components/lcn/core.py:
```diff
@@ -0,0 +1,55 @@
+CONF_TRANSITION = 'transition'
+CONF_DIMMABLE = 'dimmable'
+CONF_CONNECTIONS = 'connections'
+CONF_ADDRESS = 'address'
```
Already available in `const.py`.

alengwenus replied:
Replaced.

fabaff commented on homeassistant/components/lcn/core.py:
```diff
@@ -0,0 +1,55 @@
+"""Core dependencies for LCN component."""
+import re
+import voluptuous as vol
```
Import should be grouped.

alengwenus replied:
Done.

fabaff commented on homeassistant/components/lcn/__init__.py:
```diff
@@ -0,0 +1,118 @@
+            _LOGGER.info('LCN connected to "{:s}"'.format(connection_name))
+        except TimeoutError:
+            _LOGGER.error('Connection to PCHK server "{:s}" failed.'.format(
+                connection_name))
```
If something is wrong return immediately.

fabaff commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+from homeassistant.const import CONF_NAME
+import homeassistant.helpers.config_validation as cv
+
+from ..lcn import LcnDevice
```
Use `homeassistant.com...` instead of `..lcn`.

fabaff reviewed (commented):
Just some comments to get the review started.

alengwenus committed:
Corrected pre-review comments

alengwenus commented:
I've made the above changes. Could you re-review, please @fabaff

alengwenus committed:
Fixed dimming behaviour in combination with transitions for lcn lights

alengwenus commented:
@fabaff :  Is there anything I need to do to move this on? Did I miss anything? Please advise.

pvizeli reviewed (changes requested):
Please make a simple lcn.py as a component. I see no reason to split that two files `__init__` and `core`.

alengwenus commented:
@pvizeli The reason is that beyond this PR I already implemented a lot of more functionality (approx. 1600 LOC), especially additional constants, service calls, validators, ... which are necessary for the LCN system. To keep this first PR simple, I just removed the code but kept the module structure as it will be necessary with furure PRs.
If you really insist, I could combine everything in one lcn.py. But it will be necessary to split it up later again.

alengwenus committed:
Removed unused logger

alengwenus commented:
@pvizeli Could you please advise if I may keep the current module structure (see my comments above).

alengwenus committed:
Combined __init__.py and core.py into lcn.py component. Bumped to pypck==0.5.6

alengwenus commented:
Thanks for reviewing. I implemented all requested changes @pvizeli , @fabaff . Would you recheck, please. I'd really want to push this foreward.

alengwenus committed:
Fixed .coveragerc

alengwenus committed:
Bumped to pypck==0.5.7

alengwenus committed:
Bump to pypck==0.5.8

alengwenus committed:
Fixed requirements_all.txt

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,168 @@
+        connections = hass.data[DATA_LCN][CONF_CONNECTIONS]
+        connection = get_connection(connections, connection_id)
+
+        self.address_connection = connection.get_address_conn(addr)
```
Pass in the correct connection or `address_connection` instead. We don't want side effects in init.

alengwenus replied:
Ok, moved gathering of `address_connection` to `async_setup_platform`.

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,168 @@
+        connection = get_connection(connections, connection_id)
+
+        self.address_connection = connection.get_address_conn(addr)
+        self.address_connection.register_for_inputs(
```
What does this do? Update callbacks should be registered in entity coroutine method `async_added_to_hass`.

alengwenus replied:
This registers a callback on incoming message objects. Moved to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+
+DEPENDENCIES = ['lcn']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
```
Move the device config to the component so that all config is under the component. See eg the rainmachine component.

Load the platform via our discovery helper.

alengwenus replied:
Nice! Didn't even know this is possible. I like that.

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,168 @@
+            await hass.async_create_task(connection.async_connect())
+            connections.append(connection)
+            _LOGGER.info('LCN connected to "{:s}"'.format(connection_name))
+        except TimeoutError:
```
How long is the timeout? We're potentially holding up home assistant from starting until this times out, so it shouldn't be too long. Not more than 15-20 seconds.

alengwenus replied:
15 seconds are absolutely sufficient. Passed it as argument to `async_connect` method.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+class LcnOutputLight(LcnDevice, Light):
+    """Representation of a LCN light for output ports."""
+
+    def __init__(self, hass, config):
```
See above.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+
+    def __init__(self, hass, config):
+        """Initialize the LCN light."""
+        LcnDevice.__init__(self, hass, config)
```
```py
super().__init__(config, connection)
```

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+        LcnDevice.__init__(self, hass, config)
+
+        self.output = self.pypck.lcn_defs.OutputPort[
+            config[CONF_OUTPUT].upper()]
```
All config modifications should be made in the config schema.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+            config[CONF_OUTPUT].upper()]
+
+        self._transition = self.pypck.lcn_defs.time_to_ramp_value(
+            config[CONF_TRANSITION] * 1000)
```
See above.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+        self._is_on = None
+        self._is_dimming_to_zero = False
+
+        self.hass.async_create_task(
```
Move this to `async_added_to_hass`.

MartinHjelmare reviewed (changes requested):
Also set `should_poll` to `False` on the entity.

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,168 @@
+        # use 'pchk' as default connection_id (or add a numeric suffix if
+        # pchk' is already in use
+        connection_name = conf_connection[CONF_NAME]
+        if connection_name == DEFAULT_NAME:
```
Eg `has_all_unique_files` in the mysensors component.

alengwenus replied:
Connection name is validated in config schema using `has_unique_connection_names` function.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,118 @@
+    """Set up the LCN light platform."""
+    device = LcnOutputLight(hass, config)
+    async_add_entities([device])
+    return True
```
Remove this. Nothing is checking this return value.

alengwenus committed:
Moved unique generation of connection names to config schema's validator

alengwenus committed:
Minor changes due to review comments.
Bump to pypck==0.5.9.

alengwenus committed:
Address_connection is passed into LcnDevice

alengwenus committed:
Set should_poll property on LcnDevice to return False

alengwenus committed:
Moved platform config validation to component. Load platform using discovery helper

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,202 @@
+LIGHTS_SCHEMA = vol.Schema({
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_ADDRESS): is_address,
+    vol.Required(CONF_OUTPUT): in_upper(OUTPUT_PORTS),
```
We don't need a custom validator.
```py
vol.All(vol.In(OUTPUT_PORTS), vol.Upper)
```

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,202 @@
+
+    hass.data[DATA_LCN][CONF_CONNECTIONS] = connections
+
+    load_platform(hass, 'light', DOMAIN, config[DOMAIN][CONF_LIGHTS], config)
```
Create a task and use `async_load_platform`.
```py
hass.async_create_task(
    async_load_platform(...))
```

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,119 @@
+    def input_received(self, input_obj):
+        """Set light state when LCN input object (command) is received."""
+        if isinstance(input_obj, self.pypck.inputs.ModStatusOutput):
+            if input_obj.get_output_id() == self.output.value:
```
If we invert the two checks, combine them and return if true, we can outdent below. Ie use a guard clause.

MartinHjelmare commented on homeassistant/components/light/lcn.py:
```diff
@@ -0,0 +1,119 @@
+
+        self.address_connection.dim_output(self.output.value, percent,
+                                           transition)
+        await self.async_update_ha_state()
```
Is there no feedback via callback after calling `address_connection.dim_output`? If we would get feedback within 2 seconds, we don't need to adjust state here.

alengwenus replied:
Usually there is a feedback via a status message which is handled by the callback method. Unfortunately, it sometimes happen (due to heavy hardware bus traffic) that it does not arrive within 2 seconds. The underlying library will handle the retransmission attempt of the dimming command every 3 seconds (up to 3 attempts) if no acknowledge is received from the hardware module.

MartinHjelmare reviewed (commented):
Nice! Some more comments and questions.

alengwenus committed:
Furtehr changes due to the review

MartinHjelmare commented on homeassistant/components/lcn.py:
```diff
@@ -0,0 +1,200 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_CONNECTIONS): vol.All(
+            cv.ensure_list, has_unique_connection_names, [CONNECTION_SCHEMA]),
+        vol.Optional(CONF_LIGHTS, default=[]): vol.All(
```
Maybe we should make this required since the component without lights is useless?

alengwenus committed:
Light configuration is set required as there are no other platforms up to now

MartinHjelmare reviewed (approved):
Great!

alengwenus commented:
Thanks a lot @MartinHjelmare! Amazing work!
